### ISSUE TITLE
Floating scroll bars

### ISSUE THREAD

**Issue by Hubro** [neovide/neovide#1304]
**Is your feature request related to a problem? Please describe.**

One issue I've had with Vim since I first started using it over 10 years ago (although admittedly it's a small issue) is the fact that I have no good sense of *where* I am in a document, especially while I'm scrolling up and down.

As I press <kbd>Ctrl</kbd><kbd>u</kbd> and <kbd>Ctrl</kbd><kbd>d</kbd>, the text on the screen just changes. There is absolutely no indication how far I scrolled or how fast I am traversing the document. This is especially true when pair programming and the person watching doesn't even know if I'm scrolling up or down.

(This situation is somewhat improved by smooth scrolling, but that really only works well when scrolling short distances with the scroll wheel or <kbd>Ctrl</kbd><kbd>y</kbd>/<kbd>Ctrl</kbd><kbd>e</kbd>.)

**Describe the solution you'd like**

A floating scroll bar!

I am only suggesting something super simple. It would only be a semi-transparent indicator that floats above the buffer temporarily as the user scrolls. The height of the indicator compared to the height of the window would be the same as the height of the window compared to the height of the document. This gives an excellent indication where the user is located in the document.

The file explorer in VSCode is an excellent example of what I mean:

![image](https://user-images.githubusercontent.com/597206/167254893-3066cf57-79ea-4277-85f9-8939f0430cc9.png) ![image](https://user-images.githubusercontent.com/597206/167254923-5f33df25-217f-4002-9fe2-9daee2444eb6.png)

When scrolling here, you instantly know how many lines are left below the current window position, and you also get a great sense of how *fast* you are traversing the document while scrolling.

Since this is a floating, semi-transparent indicator, there should be no problem drawing this on top of any window without causing layout issues.

Since Neovide is hardware accelerated, the indicator could also be smoothly animated, for an extra nice touch.

The color of the indicator could be very simply decided based on the value of `'background'`. If `'background'` is `'light'` the indicator would be black. If `'background'` is `'dark'`, the indicator would be white. In both cases the indicator opacity would be something like 0.3. This would make it render nicely over pretty much any color scheme.

For the MVP, I think the scroll bar should have 2 settings:

- `g:neovide_scrollbar_enabled`
  - `v:true` - Enable the scrollbar (default)
  - `v:false` - Disable the scrollbar
- `g:neovide_scrollbar_mode`
  - `'scroll'` - Only display the scrollbar while scrolling (default)
  - `'always_on'` - Always display the indicator

To implement this, Neovide would only need the following from nvim:
- The window height
- The number of lines in the buffer displayed in the window
- An event whenever the window is scrolled (https://neovim.io/doc/user/autocmd.html#WinScrolled)
- An event whenever code is folded/unfolded (This doesn't currently exist, but is on the 0.8 milestone: https://github.com/neovim/neovim/issues/8538)
  
  For this reason, a scrollbar MVP should probably ignore folds completely.

**Describe alternatives you've considered**

The plugin https://github.com/dstein64/nvim-scrollview implements exactly this, using floating windows. However, since it's only a regular plugin, it's very limited. The scroll bar has to be one character wide, it can't be any thinner. It also can't be animated. It also doesn't work over floating windows because of some nvim limitation with stacking floating windows.

That said, I still use nvim-scrollview every day, it's awesome:

![image](https://user-images.githubusercontent.com/597206/167255281-7e1ab7a4-fb09-4737-bad6-d885d1e55d77.png)

I think Neovide, since it's a graphical UI, could implement this much, much better.

**Additional context**

Possible future enhancements:

- Customization of scrollbar width, border rounding, color and transparency
- Click-and-drag to scroll. This actually already works with `nvim-scrollview`, but only in the terminal. It doesn't work in Neovide.

**Comment by Kethku** [comment]
This is handled extremely well by https://github.com/lewis6991/satellite.nvim which I use in my personal config. It doesn't support the auto hiding or click to drag afaik but I see no reason why it cant. Closing in favor of plugin support

**Comment by 01mf02** [comment]
I would like to ask you, @Kethku, to reopen this feature request. I believe that this feature probably cannot be properly filled by a plugin, because a proper scrollbar (like it is integrated in gvim and which I totally love) can be moved by single pixels, whereas `satellite.nvim` as well as `nvim-scrollview` can be only moved by `line_height` pixels at a time. That means that native scrollbars achieve a much smoother movement than scrollbars implemented by vim plugins.
(Furthermore, inside neovide, I was not able to drag the scrollbars in neither `satellite.nvim` nor `nvim-scrollview` with the mouse at all. In neovim, that works fine, though.)

A native scroll bar would be *the* killer feature to convince me to switch to neovide. No other vim-based editor (apart from gvim) has been able to convince me to switch, for the lack of exactly this feature.

**Comment by Hubro** [comment]
Not to mention that no plugin can match the quality and performance of a native GUI.

I'm hopeful that Neovide will provide a scrollbar (and maybe a minimap like in vscode) and implementations for `vim.notify`, `vim.ui.input`, `vim.ui.select` in the future.

**Comment by mawkler** [comment]
It would be cool if the native scrollbar also had a nice fade-out animation and became invisible when the window isn't scrolled, was quite a bit thinner than the width of a column, and had a slight transparency

**Comment by imranZERO** [comment]
@Kethku scrollbars are a fundamentally GUI feature. No TUI plugin can come close to replicating the feel and smoothness of a proper scrollbar.Dragging the scrollbar with the mouse is the easiest way to explore a large file for me.

 Since neovide draws everything on it's own, it should be cleaner to implement it natively, instead of fussing around with different platform APIs.

Hoping to see this implemented.

**Comment by mawkler** [comment]
@imranZERO

> Dragging the scrollbar with the mouse

Could you elaborate on the word "mouse" please. What is that?

**Comment by fredizzimo** [comment]
Once this is merged https://github.com/neovim/neovim/pull/32535, you should be able to use plugins like [`nvim-scrollview`](https://github.com/dstein64/nvim-scrollview), with proper mouse support.

See, https://github.com/neovide/neovide/issues/2226

**Comment by imranZERO** [comment]
> [@imranZERO](https://github.com/imranZERO)
> 
> > Dragging the scrollbar with the mouse
> 
> Could you elaborate on the word "mouse" please. What is that?

It's the best way to interact with a GUI window.

**Comment by 01mf02** [comment]
> Once this is merged [neovim/neovim#32535](https://github.com/neovim/neovim/pull/32535) on the Neovim side, you should be able to use plugins like [`nvim-scrollview`](https://github.com/dstein64/nvim-scrollview), with proper mouse support.

@fredizzimo, this sounds interesting. Could you shortly elaborate what you mean by "proper mouse support"? Does that mean that a scrollbar can be moved with pixel precision in the future?
